### PR TITLE
Hotfix: Adding explicit --nodes for Slurm 26.05 to solve PartitionNodeLimit bug

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# Hotfix: Adding explicit --nodes for Slurm 26.05 to solve PartitionNodeLimit bug

*This PR description was generated with Claude Opus 4.6 (Kiro IDE) and reviewed by Terrence McGuinness, OMD/EMC.*

## Summary

Slurm 26.05 on Gaea C6 now rejects jobs that specify `--ntasks` without an explicit `--nodes` directive. Without it, Slurm calculates a worst-case node range that can exceed the partition's `MaxNodes` setting (512 for c5/c6 batch partitions), causing the scheduler to reject the job with `PartitionNodeLimit`.

This PR injects `--nodes` via Rocoto's `<native>` scheduler flags for Gaea only, preserving the existing `--ntasks`-only submission shape on all other platforms.

## What changed

| File | Change |
|------|--------|
| `dev/workflow/setup_buildxml.py` | Added `import math`. In `get_task_spec()`: when the host declares `CPUS_PER_NODE` and the scheduler is Slurm, compute `nodes = ceil(cores / cpus_per_node)` and append `--nodes=N` to the native flags. In `get_host_specs()`: pass through `CPUS_PER_NODE` from the host YAML (defaults to 0 = opt-out). |
| `dev/workflow/hosts/gaeac6.yaml` | Added `CPUS_PER_NODE: 192` (c6 nodes, per Gaea admin guidance). |

## Design approach

The Gaea admin's guidance says to add `--nodes` **alongside** the existing `--ntasks` — not replace one with the other. An earlier iteration of this fix modified `rocoto.py` to force the `<nodes>N:ppn=P:tpp=T</nodes>` rendering for all Slurm hosts. That changed the sbatch shape from `--ntasks=N` to `--nodes=N --tasks-per-node=P` on every platform, which broke Ursa (8 nodes requested for a single-node 8-core build because `CPUS_PER_NODE` defaulted incorrectly).

The final approach leaves `rocoto.py` and the `<cores>` rendering untouched. Instead, `--nodes` is appended to the `<native>` tag — which Rocoto passes through verbatim to `#SBATCH`. This is opt-in per host via `CPUS_PER_NODE` in the host YAML:

- **Hosts with `CPUS_PER_NODE` set** (Gaea C6): get `--nodes=ceil(cores/192)` appended to native.
- **Hosts without it** (Ursa, Hera, Hercules, Orion, etc.): completely unaffected. Their sbatch scripts remain `--ntasks=N` only, identical to develop.

## Example output

**Gaea C6** — `upp` build (8 cores, `CPUS_PER_NODE: 192`):
```
#SBATCH --ntasks=8
#SBATCH --export=NONE --clusters=es --nodes=1
```

**Ursa** — `upp` build (8 cores, no `CPUS_PER_NODE`):
```
#SBATCH --ntasks=8
#SBATCH --export=NONE
```

## Tested

- **Gaea C6** (`eslogin_c6`, Slurm 26.05): `build_all.sh` all jobs submitted and completed successfully.
No `PartitionNodeLimit`.

```
Terry.McGuinness (gaea64) sorc (hotfix/gaea_slurm_update_build) $ ./build_all.sh -A drsa-precip3 -c all
Sourcing global-workflow modules ...
Running "module reset". Resetting modules to system default. 
Generating build.xml for building global-workflow programs ...
Building on compute nodes using account: drsa-precip3
-----------------------------------
| System             | Status     |
-----------------------------------
| gsi_enkf           | SUCCEEDED  |
| gcafs_model        | SUCCEEDED  |
| nexus              | SUCCEEDED  |
| gfs_model          | SUCCEEDED  |
| gfs_utils          | SUCCEEDED  |
| gefs_model         | SUCCEEDED  |
| gefs_ww3_prepost   | SUCCEEDED  |
| gdas               | SUCCEEDED  |
| ufs_utils          | SUCCEEDED  |
| gfs_ww3prepost     | SUCCEEDED  |
| gsi_utils          | SUCCEEDED  |
| gsi_monitor        | SUCCEEDED  |
| sfs_model          | SUCCEEDED  |
| upp                | SUCCEEDED  |
-----------------------------------
All builds completed successfully!
```
- **Ursa** (`u1-compute`): `build_all.sh` — all jobs submitted and completed successfully.
No change in submission shape vs develop.
image

- **Hercules**
```
mterry (hercules-login-1) sorc (hotfix/gaea_slurm_update_build) $ ./build_all.sh -A $HPC_ACCOUNT -c all
Sourcing global-workflow modules ...
Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: /work2/noaa/global/mterry/global-workflow_forked/modulefiles
Generating build.xml for building global-workflow programs ...
Building on compute nodes using account: fv3-cpu
-----------------------------------
| System             | Status     |
-----------------------------------
| upp                | SUCCEEDED  |
| gsi_utils          | SUCCEEDED  |
| gefs_ww3_prepost   | SUCCEEDED  |
| ufs_utils          | SUCCEEDED  |
| gcafs_model        | RUNNING    |
| gfs_ww3prepost     | SUCCEEDED  |
| gdas               | RUNNING    |
| gsi_enkf           | SUCCEEDED  |
| gfs_utils          | SUCCEEDED  |
| sfs_model          | RUNNING    |
| gefs_model         | RUNNING    |
| gsi_monitor        | SUCCEEDED  |
| nexus              | SUCCEEDED  |
| gfs_model          | RUNNING    |
-----------------------------------
```

## Context

Per Gaea C6 admin notice (August 2026):

> Slurm 26.05 now requires jobs that use `--ntasks` without an explicit `--nodes` directive to specify the node count directly. Without `--nodes`, Slurm calculates a worst-case node range that can exceed the partition's maximum node limit, causing the scheduler to reject the job with `PartitionNodeLimit`.

Calculation per the admin's guidance: for c6 (192 CPUs/node), divide `--ntasks` by 192 and round up. For the current build workload (max 8 cores after the Gaea host override), `ceil(8/192) = 1`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] New and existing tests pass with my changes
- [x] My changes generate no new warnings
- [x] Tested on Gaea C6 (target platform)
- [x] Tested on Ursa (regression check — no change in behavior)

### CODE REVIEW

File: dev/workflow/setup_buildxml.py [hunk 1]
```diff
@@ -77,9 +78,16 @@ def get_task_spec(task_name: str, task_spec: Dict, host_spec: Dict) -> Dict:
...
+
+    # Slurm 26.05+ requires explicit --nodes. Compute from ntasks and
+    # cpus_per_node (carried in host_spec, defaults to ntasks fitting one node).
+    ntasks = task_spec.cores
+    cpus_per_node = host_spec.get("cpus_per_node", ntasks)
+    nodes = math.ceil(ntasks / cpus_per_node)
+
+    task_dict.resources.nodes = nodes
+    task_dict.resources.ntasks = ntasks
+    task_dict.resources.ppn = min(ntasks, cpus_per_node)
```
[the remark this answers is not in the document]
Comment (TerrenceMcGuinness-NOAA) [reply dev/workflow/setup_buildxml.py]: This looks like a good catch of that corner case.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Build fails on **Ursa** right off with update.
Smoke tests needs be tested on all platforms before re-opening PR.

```
08/13/26 22:14:18 UTC :: build.xml :: Cycle 190001010000, Task gfs_ww3prepost, jobid=19327527, in state DEAD (OUT_OF_MEMORY), ran for 128.0 seconds, exit status=125, try=1 (of 1)
08/13/26 22:14:18 UTC :: build.xml :: Cycle 190001010000, Task gefs_ww3_prepost, jobid=19327530, in state DEAD (OUT_OF_MEMORY), ran for 129.0 seconds, exit status=125, try=1 (of 1)
08/13/26 22:14:18 UTC :: build.xml :: Cycle 190001010000, Task upp, jobid=19327533, in state DEAD (OUT_OF_MEMORY), ran for 133.0 seconds, exit status=125, try=1 (of 1)
-----------------------------------
| System             | Status     |
-----------------------------------
| upp                | DEAD       |
| gsi_utils          | SUCCEEDED  |
| gefs_ww3_prepost   | DEAD       |
| ufs_utils          | SUCCEEDED  |
| gcafs_model        | RUNNING    |
| gfs_ww3prepost     | DEAD       |
| gdas               | QUEUED     |
| gsi_enkf           | RUNNING    |
| gfs_utils          | SUCCEEDED  |
| sfs_model          | RUNNING    |
| gefs_model         | RUNNING    |
| gsi_monitor        | SUCCEEDED  |
| nexus              | SUCCEEDED  |
| gfs_model          | RUNNING    |
-----------------------------------
FATAL ERROR: The following builds failed, see log files for details:
	upp: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_5212_7d104183_67693/global-workflow/sorc/logs/upp.log
	gefs_ww3_prepost: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_5212_7d104183_67693/global-workflow/sorc/logs/gefs_ww3_prepost.log
	gfs_ww3prepost: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_5212_7d104183_67693/global-workflow/sorc/logs/gfs_ww3prepost.log
```

**Comment by DavidHuber-NOAA** [comment]
The Hercules failures seem to be in step with the power failure on that system. Reviewing the successful Hercules tests up to the point of failure, everything seemed to work nominally. Merging.

Noting that the errors on Hercules show that there is a malformed call to rocoto in the case of an `UNAVAILABLE` state reported by rocoto. This should be fixed, but does not need to hold up this PR.

**Review by DavidHuber-NOAA — APPROVED** [review]
